File: src/main/java/io/shiftleft/fuzzyc2cpg/output/inmemory/OutputModuleFactory.java
package io.shiftleft.fuzzyc2cpg.output.inmemory;

import io.shiftleft.codepropertygraph.Cpg;
import io.shiftleft.fuzzyc2cpg.output.CpgOutputModule;
import io.shiftleft.fuzzyc2cpg.output.CpgOutputModuleFactory;

public class OutputModuleFactory implements CpgOutputModuleFactory {

  private OutputModule outputModule;

  @Override
  public CpgOutputModule create() {
    synchronized (this) {
      if (outputModule == null) {
        outputModule = new OutputModule();
      }
    }
    return outputModule;
  }

  /**
   * An internal representation of the graph.
   *
   * @return the internally constructed graph
   */
  public Cpg getInternalGraph() {
    return outputModule.getInternalGraph();
  }

  @Override
  public void persist() {
    if (outputModule != null) {
      outputModule.persist();
    }
  }
}

File: src/main/scala/io/shiftleft/fuzzyc2cpg/output/inmemory/OutputModuleFactory.scala
package io.shiftleft.fuzzyc2cpg.output.inmemory

import io.shiftleft.codepropertygraph.Cpg
import io.shiftleft.fuzzyc2cpg.output.CpgOutputModuleFactory

class OutputModuleFactory extends CpgOutputModuleFactory {
  private var outputModule: OutputModule = _

  override def create(): OutputModule = {
    if (outputModule == null) outputModule = new OutputModule()
    outputModule
  }

  /**
    * An internal representation of the graph.
    *
    * @return the internally constructed graph
    */
  def getInternalGraph: Cpg = outputModule.getInternalGraph

  override def persist(): Unit = if (outputModule != null) outputModule.persist()
}
